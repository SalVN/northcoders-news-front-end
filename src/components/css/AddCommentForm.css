
.comment-form-heading {
   font-weight: bold;
   color: #FC84A0;
}

.comment-form {
    text-align: center;
}

.box-style-comment {
    background-color: #FC84A0;
}

.box-style-comment-small {
    height: 70px;
}

.comment-form-image {
    height: 36px;
    width: 36px;
    border-radius: 20px;
}

.comment-form-small {
    border-radius: 10px;
    width: 98%;
    background-color: white;
    height: 36px;
    padding: 5px;
    font-size: 14px;
    line-height: 20px;
    border: 1px solid black;
}
.add-comment-textarea {
    resize: vertical;
    width: 98%;
    min-height: 100px;
    max-height: 300px;
    font-size: 14px;
    padding: 5px;
    margin-top: -28px;
    border-radius: 10px;
    border: 1px solid black;
}

.button-div {
    text-align: right !important;
    margin-top: 10px;
}

.add-comment-submit {
    background-color: #c1002e;
    color: white;
    border-radius: 20px;
    border: 0px;
    margin-top: 0px;
}

.characters-left {
    margin-right: 20px;
    font-weight: 400;
}