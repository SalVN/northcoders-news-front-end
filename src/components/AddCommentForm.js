import React from 'react';
import PropTypes from 'prop-types';

import './css/AddCommentForm.css';

class AddCommentForm extends React.Component {
    constructor(props) {
        super(props);
        this.state = {
            text: ''
        };
        this.handleChange = this.handleChange.bind(this);
    }
    render() {
        return (
            <div>
                {this.props.showForm
                    ?
                    <div className='box box-style box-style-comment box-style-comment'>
                        <article className='media'>
                            <div className='media-left'>
                                <img className='comment-form-image' src={this.props.user.avatar_url} />
                            </div>
                            <div className='media-content'>
                                <div className='content'>
                                    <form onSubmit={this.props.handleSubmit}>
                                        <label className='comment-form-heading' htmlFor="comment-input">Join the conversation:</label>
                                        <br />
                                        <div className='comment-form' onBlur={this.props.toggleForm.bind(this, this.state.text)}>
                                            <textarea className='add-comment-textarea' onChange={this.handleChange} id="comment-input" defaultValue={this.state.text} placeholder='Join the conversation' />
                                        </div>
                                        <div className='button-div'>
                                            <button className='button add-comment-submit' type='submit'>Submit</button>
                                        </div>
                                    </form>
                                </div>
                            </div>
                        </article>
                    </div>
                    : <div className='box box-style box-style-comment box-style-comment-small'>
                        <article className='media'>
                            <div className='media-left'>
                                <img className='comment-form-image' src={this.props.user.avatar_url} />
                            </div>
                            <div className='media-content'>
                                <div className='content'>
                                    <div onClick={this.props.toggleForm.bind(this, null)} className='comment-form-small'>
                                        <p>Join the conversation</p>
                                    </div>
                                </div>
                            </div>
                        </article>
                    </div>
                }
            </div>
        );
    }
    handleChange(e) {
        this.setState({
            text: e.target.value
        });
    }
}

AddCommentForm.propTypes = {
    id: PropTypes.string.isRequired,
    addComment: PropTypes.func.isRequired,
    toggleForm: PropTypes.func.isRequired,
    user: PropTypes.object.isRequired,
    showForm: PropTypes.bool.isRequired,
    handleSubmit: PropTypes.func.isRequired
};

export default AddCommentForm;