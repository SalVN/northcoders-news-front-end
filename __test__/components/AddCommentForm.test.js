import React from 'react';
import { shallow, mount } from 'enzyme';
import renderer from 'react-test-renderer';
import { MemoryRouter } from 'react-router-dom';
import sinon from 'sinon';

import AddCommentForm from '../../src/components/AddCommentForm';

describe('AddCommentForm', () => {
    const user = {
        _id: '59b1b18b327cce1fb043bdb1',
        username: 'northcoder',
        name: 'Awesome Northcoder',
        avatar_url: 'https://avatars3.githubusercontent.com/u/6791502?v=3&s=200',
        __v: 0
    };
    it('is a function', () => {
        expect(typeof AddCommentForm).toEqual('function');
    });

    it('renders', () => {
        const wrapper = shallow(<AddCommentForm
            user={user}
            id={user._id}
            showForm={true}
            handleSubmit={x => x}
            toggleForm={x => x}
        />);
        expect(wrapper.children().length).toEqual(1);
    });

    it('renders correctly', () => {
        const tree = renderer.create(
            <MemoryRouter>
                <AddCommentForm
                    user={user}
                    id={user._id}
                    showForm={false}
                    handleSubmit={x => x}
                    toggleForm={x => x}
                />
            </MemoryRouter>
        ).toJSON();
        expect(tree).toMatchSnapshot();
    });

    it('will not show the form if showForm is false', () => {
        const wrapper = shallow(<AddCommentForm
            user={user}
            id={user._id}
            showForm={false}
            handleSubmit={x => x}
            toggleForm={x => x}
        />);
        expect(wrapper.find('form').length).toEqual(0);
    });

    it('will shows the form if showForm is true', () => {
        const wrapper = shallow(<AddCommentForm
            user={user}
            id={user._id}
            showForm={true}
            handleSubmit={x => x}
            toggleForm={x => x}
        />);
        expect(wrapper.find('form').length).toEqual(1);
    });

    it('toggles the form if the div is clicked', () => {
        const spy = sinon.stub();
        const wrapper = shallow(<AddCommentForm
            user={user}
            id={user._id}
            showForm={false}
            handleSubmit={x => x}
            toggleForm={spy}
        />);
        expect(spy.called).toBe(false);
        wrapper.find('.comment-form-small').simulate('click', { preventDefault() { } });
        expect(spy.callCount).toBe(1);
    });


    it('toggles the form if the form is blurred', () => {
        const spy = sinon.stub();
        const wrapper = mount(<AddCommentForm
            user={user}
            id={user._id}
            showForm={true}
            handleSubmit={x => x}
            toggleForm={spy}
        />);
        expect(spy.called).toBe(false);
        wrapper.find('textarea').simulate('click');
        wrapper.find('.comment-form').simulate('blur', { preventDefault() { } });
        expect(spy.callCount).toBe(1);
    });

    it('submits the form onSubmit', () => {
        const spy = sinon.stub();
        const wrapper = shallow(<AddCommentForm
            user={user}
            id={user._id}
            showForm={true}
            handleSubmit={spy}
            toggleForm={x => x}
        />);
        wrapper.setState({ 'text': 'foo' });
        wrapper.find('form').simulate('submit');
        expect(spy.callCount).toBe(1);
    });

    it('will disable the form submit if there is no text contained in the box', () => {
        const wrapper = shallow(<AddCommentForm
            user={user}
            id={user._id}
            showForm={true}
            handleSubmit={x => x}
            toggleForm={x => x}
        />);
        expect(wrapper.find('.add-comment-submit').node.props.disabled).toBe(true);

        wrapper.setState({ 'text': 'foo' });
        expect(wrapper.find('.add-comment-submit').node.props.disabled).toBe(false);
    });

    it('displays the characters left', () => {
        const wrapper = shallow(<AddCommentForm
            user={user}
            id={user._id}
            showForm={true}
            handleSubmit={x => x}
            toggleForm={x => x}
        />);
        expect(wrapper.find('.characters-left').node.props.children).toBe(1000);
        wrapper.setState({ 'text': 'hello world' });
        expect(wrapper.find('.characters-left').node.props.children).toBe(989);
    });

    it('changes the state if the form is changed', () => {
        const wrapper = mount(<AddCommentForm
            user={user}
            id={user._id}
            showForm={true}
            handleSubmit={x => x}
            toggleForm={x => x}
        />);
        wrapper.find('textarea').simulate('change', { target: { value: 'text' } });
        expect(wrapper.state('text')).toBe('text');
    });

});