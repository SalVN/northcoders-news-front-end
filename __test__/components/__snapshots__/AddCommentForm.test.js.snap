// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`AddCommentForm renders correctly 1`] = `
<div>
  <div
    className="box box-style box-style-comment box-style-comment-small"
  >
    <article
      className="media"
    >
      <div
        className="media-left"
      >
        <img
          className="comment-form-image"
          src="https://avatars3.githubusercontent.com/u/6791502?v=3&s=200"
        />
      </div>
      <div
        className="media-content"
      >
        <div
          className="content"
        >
          <div
            className="comment-form-small"
            onClick={[Function]}
          >
            <p>
              Join the conversation
            </p>
          </div>
        </div>
      </div>
    </article>
  </div>
</div>
`;
